##### Analyst Test -Part2:
    1. Determine if the task is not completed
    2. Define who the task is designed to
    3. How the workload of the not completed tasks divided among the care coordinators
    4. The average time to complete each category of task for the tasks created in 2018


#### Workload of uncompleted tasks among Care Coordinators
#### 1.a Determine if the task is not completed

In [60]:
# Import pandas packages
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns

In [61]:
# Upload Analyst_Test_Tasks.csv
tasks = pd.read_csv('Analyst_Test_Tasks.csv')
tasks.head()

Patient drchrono ID             Category    Status  Priority Assigned To  \
0             60094576  Specimen Collection  Complete        20  MOXXIadmin   
1             59946176           Phone Call  Complete        20  MOXXIadmin   
2             59740292           Phone Call  Complete        20  MOXXIadmin   
3             60121538           Phone Call  Complete        20  MOXXIadmin   
4             60121538                  NaN  Complete        20  MOXXIadmin   

            Created on    Create by           Updated on  \
0  2016-07-22T14:14:20  Neilsolomon  2018-03-28T21:14:32   
1  2016-08-25T20:07:00        nsome  2018-03-28T21:14:32   
2  2016-08-29T18:23:10        nsome  2018-03-28T21:14:32   
3  2016-09-02T15:08:08        nsome  2018-03-28T21:14:32   
4  2016-09-06T08:54:44        nsome  2018-03-28T21:14:32   

             Next Appt Date  HC20 Status          HC20 Substatus  \
0                       NaN          NaN                     NaN   
1  2018-07-26T17:00:00.000Z     Enrolled               Scheduled   
2  2018-03-30T15:30:00.000Z  Disenrolled  Ineligible - insurance   
3                       NaN  Disenrolled  Ineligible - insurance   
4                       NaN  Disenrolled  Ineligible - insurance   

  Stratification level Last Appointment Date  
0                  NaN   2016-08-05 21:30:00  
1                 High   2018-06-20 21:30:00  
2                  NaN   2018-02-21 19:15:00  
3                  NaN   2018-03-13 16:55:00  
4                  NaN   2018-03-13 16:55:00

In [62]:
# Overview the info on tasks set
tasks.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3659 entries, 0 to 3658
Data columns (total 13 columns):
Patient drchrono ID      3659 non-null int64
Category                 3445 non-null object
Status                   3659 non-null object
Priority                 3659 non-null int64
Assigned To              3598 non-null object
Created on               3659 non-null object
Create by                2961 non-null object
Updated on               3659 non-null object
Next Appt Date           2411 non-null object
HC20 Status              3646 non-null object
HC20 Substatus           3646 non-null object
Stratification level     621 non-null object
Last Appointment Date    3659 non-null object
dtypes: int64(2), object(11)
memory usage: 371.7+ KB


In [63]:
# Check for null value (NAN)
tasks.isnull().sum()

Patient drchrono ID         0
Category                  214
Status                      0
Priority                    0
Assigned To                61
Created on                  0
Create by                 698
Updated on                  0
Next Appt Date           1248
HC20 Status                13
HC20 Substatus             13
Stratification level     3038
Last Appointment Date       0
dtype: int64

In [64]:
tasks.drop(['Stratification level', 'Next Appt Date','Create by'],axis =1, inplace=True)

In [65]:
tasks.dropna(inplace=True)

In [66]:
tasks.isnull().sum()

Patient drchrono ID      0
Category                 0
Status                   0
Priority                 0
Assigned To              0
Created on               0
Updated on               0
HC20 Status              0
HC20 Substatus           0
Last Appointment Date    0
dtype: int64

In [67]:
# Check how many categories in Status and the number of each category
tasks['Status'].value_counts()

Complete       2494
In Progress     633
Open            149
On Hold          58
Authorized       39
Back to PCP       3
Name: Status, dtype: int64

####  List of Assignees for Tasks

In [68]:
# Check how many people are assigned to modify the dataset
tasks['Assigned To'].value_counts()

MOXXIadmin          2317
suzztasharma         203
gblandon             189
marlenegarcia        163
CareCoordinators     110
pvallejo             108
aarreola              92
darzate               68
kelder                15
aalmanza              15
mdavis07              12
cmathews              11
kapostol               9
ebenons                7
HRubin                 7
epopoca                6
Neilsolomon            5
nsome                  4
stinajero              4
hanzaldua              4
ggardner               4
mduarte                4
hcolomer               3
AmandaMay              3
egragnani              2
mariadperez            2
GFedalizo              1
CMiranda               1
alexgerson             1
cristianpacheco        1
DanaLewis              1
MyrnaGonzalez          1
aglaros                1
CSobrian               1
mikejimenez            1
Name: Assigned To, dtype: int64

#### Most assignees are in Administration, followed by CareCoordinators

#### Uncompleted Tasks List

In [69]:
# Check for uncomplete task
tasks_unfinished = tasks.loc[tasks['Status'] !='Complete']
tasks_unfinished.head()

Patient drchrono ID                 Category       Status  Priority  \
262             69026616                 Referral  In Progress        20   
368             69321260  Medical Records Request      On Hold        20   
377             68793552  Medical Records Request      On Hold        20   
391             67917278  Medical Records Request  In Progress        20   
427             70296028  Medical Records Request  In Progress        40   

          Assigned To           Created on           Updated on HC20 Status  \
262          aarreola  2017-10-13T13:49:15  2018-06-28T19:27:28    Enrolled   
368  CareCoordinators  2017-11-30T16:25:15  2018-07-09T11:44:21    Enrolled   
377          cmathews  2017-12-04T13:10:20  2018-06-25T16:08:44    Enrolled   
391      suzztasharma  2017-12-06T14:17:21  2018-06-19T17:21:40    Enrolled   
427      suzztasharma  2017-12-11T20:02:23  2018-06-08T17:41:18    Enrolled   

     HC20 Substatus Last Appointment Date  
262       Scheduled   2018-05-17 16:00:00  
368  Appt completed   2018-06-01 18:00:00  
377       Scheduled   2018-07-06 20:15:00  
391       Scheduled   2018-06-08 19:15:00  
427       Scheduled   2018-07-09 20:30:00

In [70]:
# Check who weren't complete the status they were assigned to
tasks_unfinished['Assigned To'].value_counts()

suzztasharma        198
marlenegarcia       152
gblandon            132
CareCoordinators     95
aarreola             83
pvallejo             80
darzate              68
cmathews             10
aalmanza              7
kapostol              7
ebenons               6
HRubin                6
Neilsolomon           5
ggardner              4
hanzaldua             4
mduarte               4
stinajero             4
nsome                 3
epopoca               3
MOXXIadmin            3
mariadperez           2
DanaLewis             1
hcolomer              1
kelder                1
egragnani             1
mdavis07              1
MyrnaGonzalez         1
Name: Assigned To, dtype: int64

In [71]:
# Name of Care Coordinators from the Care Coordinator Names data set
CareCoordiantor_Names = pd.read_csv('Care Coordinator Names.csv')
CareCoordiantor_Names

There are 7 care coordinators:
0                       Aarreola
1                       Pvallego
2                  marlenegarcia
3                   suzztasharma
4               CareCoordinators
5                        darzate
6                       gblandon

##### Uncompleted Workload divided among Care Coordinators and the number of cases they are assigned to
1. suzztasharma        198
2. marlenegarcia       152
3. gblandon            133
4. CareCoordinators     95
5. aarreola             83
6. pvallejo             80
7. darzate              68

**** Suzztasharm has the large amount of uncompleted file, Darzate has the least uncomppleted file


#### Completed tasks

In [72]:
# Check for the Completed tasks
tasks_completed = tasks.loc[tasks['Status'] =='Complete']
tasks_completed.head()

Patient drchrono ID             Category    Status  Priority Assigned To  \
1             59946176           Phone Call  Complete        20  MOXXIadmin   
2             59740292           Phone Call  Complete        20  MOXXIadmin   
3             60121538           Phone Call  Complete        20  MOXXIadmin   
5             60794436           Phone Call  Complete        20  MOXXIadmin   
6             60121538  Specimen Collection  Complete        20  MOXXIadmin   

            Created on           Updated on  HC20 Status  \
1  2016-08-25T20:07:00  2018-03-28T21:14:32     Enrolled   
2  2016-08-29T18:23:10  2018-03-28T21:14:32  Disenrolled   
3  2016-09-02T15:08:08  2018-03-28T21:14:32  Disenrolled   
5  2016-09-14T18:30:16  2018-03-28T21:14:32  Closed lead   
6  2016-10-05T18:44:30  2018-03-28T21:14:32  Disenrolled   

           HC20 Substatus Last Appointment Date  
1               Scheduled   2018-06-20 21:30:00  
2  Ineligible - insurance   2018-02-21 19:15:00  
3  Ineligible - insurance   2018-03-13 16:55:00  
5    Ineligible - medical   1970-01-01 00:00:00  
6  Ineligible - insurance   2018-03-13 16:55:00

In [73]:
# check to see how many tasks in the data set (11 categories of tasks)
tasks_completed['Category'].value_counts()

Referral                   866
Phone Call                 592
Pharmacy Request           242
Medical Records Request    207
Lab Results                197
DME                        153
Social Services             81
Specimen Collection         56
Supplies                    47
Transportation              28
Behavioral Health           25
Name: Category, dtype: int64

In [15]:
tasks_completed.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2494 entries, 1 to 3654
Data columns (total 10 columns):
Patient drchrono ID      2494 non-null int64
Category                 2494 non-null object
Status                   2494 non-null object
Priority                 2494 non-null int64
Assigned To              2494 non-null object
Created on               2494 non-null object
Updated on               2494 non-null object
HC20 Status              2494 non-null object
HC20 Substatus           2494 non-null object
Last Appointment Date    2494 non-null object
dtypes: int64(2), object(8)
memory usage: 214.3+ KB


In [74]:
# convert to datetime column
tasks_completed['Created on']=pd.to_datetime(tasks_completed['Created on'],format='%Y-%m-%d')
tasks_completed['Updated on']=pd.to_datetime(tasks_completed['Updated on'],format='%Y-%m-%d')


/Users/thoashook/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  
/Users/thoashook/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  This is separate from the ipykernel package so we can avoid doing imports until


In [75]:
tasks_completed.head()

Patient drchrono ID             Category    Status  Priority Assigned To  \
1             59946176           Phone Call  Complete        20  MOXXIadmin   
2             59740292           Phone Call  Complete        20  MOXXIadmin   
3             60121538           Phone Call  Complete        20  MOXXIadmin   
5             60794436           Phone Call  Complete        20  MOXXIadmin   
6             60121538  Specimen Collection  Complete        20  MOXXIadmin   

           Created on          Updated on  HC20 Status  \
1 2016-08-25 20:07:00 2018-03-28 21:14:32     Enrolled   
2 2016-08-29 18:23:10 2018-03-28 21:14:32  Disenrolled   
3 2016-09-02 15:08:08 2018-03-28 21:14:32  Disenrolled   
5 2016-09-14 18:30:16 2018-03-28 21:14:32  Closed lead   
6 2016-10-05 18:44:30 2018-03-28 21:14:32  Disenrolled   

           HC20 Substatus Last Appointment Date  
1               Scheduled   2018-06-20 21:30:00  
2  Ineligible - insurance   2018-02-21 19:15:00  
3  Ineligible - insurance   2018-03-13 16:55:00  
5    Ineligible - medical   1970-01-01 00:00:00  
6  Ineligible - insurance   2018-03-13 16:55:00

In [27]:
tasked = tasks_completed[(tasks_completed['Created on'] > '2018-01-01 01:00:00') & (tasks_completed['Created on'] <= '2018-12-31 23:00:00')]
tasked.head()

Patient drchrono ID    Category    Status  Priority Assigned To  \
563             65069468  Phone Call  Complete        20  MOXXIadmin   
565             62620150         DME  Complete        20  MOXXIadmin   
567             63200518    Referral  Complete        20    gblandon   
568             63515612  Phone Call  Complete        20  MOXXIadmin   
569             70255162  Phone Call  Complete        20  MOXXIadmin   

             Created on          Updated on  HC20 Status  \
563 2018-01-02 11:22:26 2018-03-28 21:15:02  Disenrolled   
565 2018-01-02 15:34:46 2018-05-30 18:59:05     Enrolled   
567 2018-01-02 18:06:09 2018-03-28 21:15:02     Enrolled   
568 2018-01-03 13:33:59 2018-03-28 21:15:02     Enrolled   
569 2018-01-03 17:32:52 2018-03-28 21:15:02     Enrolled   

             HC20 Substatus Last Appointment Date          duration  
563  Ineligible - insurance   2018-04-04 19:50:00  85 days 09:52:36  
565          Appt completed   2018-06-22 16:00:00 148 days 03:24:19  
567               Scheduled   2018-06-20 18:45:00  85 days 03:08:53  
568               Scheduled   2018-07-02 22:50:00  84 days 07:41:03  
569               Scheduled   2018-06-28 16:30:00  84 days 03:42:10

In [28]:
tasked['duration']= tasks_completed['Updated on']- tasks_completed['Created on']

/Users/thoashook/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [46]:
tasked['duration']

563     85 days 09:52:36
565    148 days 03:24:19
567     85 days 03:08:53
568     84 days 07:41:03
569     84 days 03:42:10
              ...       
3618     0 days 00:55:39
3622     0 days 23:58:06
3624     0 days 21:21:22
3629     0 days 04:04:20
3654     0 days 00:59:24
Name: duration, Length: 2061, dtype: timedelta64[ns]

In [56]:
tasked['duration']=tasked['duration'].dt.days.astype('int16')

/Users/thoashook/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [58]:
avg_complete_cat_duration = tasked[['Category','duration']]

In [59]:
avg_complete_cat_duration.groupby('Category').mean()

duration
Category                          
Behavioral Health        40.583333
DME                      51.369128
Lab Results              12.345178
Medical Records Request  35.236559
Pharmacy Request         13.233333
Phone Call               26.100000
Referral                 48.578279
Social Services          32.402778
Specimen Collection      41.193548
Supplies                 26.911111
Transportation           27.250000

In [55]:
tasked.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2061 entries, 563 to 3654
Data columns (total 11 columns):
Patient drchrono ID      2061 non-null int64
Category                 2061 non-null object
Status                   2061 non-null object
Priority                 2061 non-null int64
Assigned To              2061 non-null object
Created on               2061 non-null datetime64[ns]
Updated on               2061 non-null datetime64[ns]
HC20 Status              2061 non-null object
HC20 Substatus           2061 non-null object
Last Appointment Date    2061 non-null object
duration                 2061 non-null timedelta64[ns]
dtypes: datetime64[ns](2), int64(2), object(6), timedelta64[ns](1)
memory usage: 193.2+ KB


In [47]:
y = [3,4,5,6,7,8]
y[:3]

[3, 4, 5]

In [ ]:
# Convert UTC datetime format to US/Pacific datetime format
#import datetime
#for x in ['Created on', 'Updated on','Last Appointment Date']:
    #if x in tasks.columns:
        #tasks[x] = pd.to_datetime(tasks[x], format="%Y-%m-%d").dt.floor("d")

In [ ]:
# Convert Date to a datetime column
#tasks_completed['Created on'] = pd.to_datetime(tasks_completed['Created on'], format='%d/%m/%y')
#tasks_completed['Updated on'] = pd.to_datetime(tasks_completed['Updated on'], format='%d/%m/%y')

In [ ]:
#tasks_completed.groupby(['Category','Duration']).sum()

In [ ]:
#tasks_completed['Duration']=tasks_completed['Duration'].map(lambda x: str(x)[:-4])
#tasks_completed['Duration']

In [ ]:
#cat_dur = tasks_completed.groupby(['Category']).agg({'Duration':['mean']})
